scripts/sysusers.sh

```diff
@@ -0,0 +1,150 @@
+#!/bin/sh
+# -*- mode: shell-script; indent-tabs-mode: true; tab-width: 4; -*-
+
+# Replacement for systemd-sysusers(8) for the use of RPM
+# Based on sysusers.generate-pre.sh from the Fedora systemd package
+
+SYSTEMDIR=/usr/lib/sysusers.d
+RUNTIMEDIR=/run/sysusers.d
+SYSTEMCONFDIR=/etc/sysusers.d
+
+ROOT=/
+REPLACE=#
+
+while test $# -gt 0 ; do
+	case "${1}" in
+		--root )
+			shift
+			ROOT=${1}
+			shift
+			;;
+		--replace )
+			shift
+			REPLACE=${1}
+			shift
+			;;
+		* )
+			break
+			;;
+	esac
+done
+
+hasuser() {
+	grep "^${1}:" "$ROOT"/etc/passwd >/dev/null
+}
+
+hasuid() {
+	cut -d: -f3 "$ROOT"/etc/passwd | grep "${1}"  >/dev/null
+}
+
+hasgroup() {
+	grep "^${1}:" "$ROOT"/etc/group >/dev/null
+}
+
+user() {
+	user="$1"
+	uid="$2"
+	desc="$3"
+	group="$4"
+	home="$5"
+	shell="$6"
+
+	[ "$desc" = '-' ] && desc=
+	{ [ "$home" = '-' ] || [ "$home" = '' ]; } && home=/
+	{ [ "$shell" = '-' ] || [ "$shell" = '' ]; } && shell=/usr/sbin/nologin
+
+	if [ "$uid" = '-' ] || [ "$uid" = '' ]; then
+		hasuser "$user" || \\
+		    useradd -R "$ROOT" -r -g "${group}" -d "${home}" -s "${shell}" -c "${desc}" "${user}" || :
+	else
+		if ! hasuser "${user}" ; then
+		    if ! hasuid "${uid}" ; then
+		        useradd -R "$ROOT" -r -u "${uid}" -g "${group}" -d "${home}" -s "${shell}" -c "${desc}" "${user}" || :
+		    else
+		        useradd -R "$ROOT" -r -g "${group}" -d "${home}" -s "${shell}" -c "${desc}" "${user}" || :
+		    fi
+		fi
+	fi
+}
+
+group() {
+	group="$1"
+	gid="$2"
+
+	if [ "$gid" = '-' ]; then
+		hasgroup "${group}" || groupadd -R "$ROOT" -f -r "${group}" || :
+	else
+		hasgroup "${group}" || groupadd -R "$ROOT" -f -g "${gid}" -r "${group}" || :
+	fi
+}
+
+usermod() {
+	user="$1"
+	group="$2"
+
+	if hasgroup "${group}" ; then
+	    usermod -R "$ROOT" -a -G "${group}" "$user" || :
+	fi
+}
+
+parse() {
+	while read -r line || [ -n "$line" ] ; do
+		{ [ "${line:0:1}" = '#' ] || [ "${line:0:1}" = ';' ]; } && continue
+		line="${line## *}"
+		[ -z "$line" ] && continue
+		eval "arr=( $line )"
+		case "${arr[0]}" in
+			('u')
+				if [[ "${arr[2]}" == *":"* ]]; then
+					user "${arr[1]}" "${arr[2]%:*}" "${arr[3]}" "${arr[2]#*:}" "${arr[4]}" "${arr[5]}"
+				else
+					group "${arr[1]}" "${arr[2]}"
+					user "${arr[1]}" "${arr[2]}" "${arr[3]}" "${arr[1]}" "${arr[4]}" "${arr[5]}"
+				fi
+				;;
+			('g')
+				group "${arr[1]}" "${arr[2]}"
+				;;
+			('m')
+				group "${arr[2]}" "-"
+				user "${arr[1]}" "-" "" "${arr[1]}" "" ""
+				usermod "${arr[1]}" "${arr[2]}"
+				;;
+		esac
+	done
+}
+
+
+# Make sure etc and the user/group files exist
+mkdir -p "$ROOT"/etc/
+[ -e "$ROOT"/etc/passwd ] || touch "$ROOT"/etc/passwd
+[ -e "$ROOT"/etc/shadow ] || touch "$ROOT"/etc/shadow
+[ -e "$ROOT"/etc/group ] || touch "$ROOT"/etc/group
```

Comment:
Why? Shouldn't things work even if those files are missing?

Reply:
While I agree `useradd` and `groupadd` just won't work. And we want rpm -q --root to just work without needing to populate the chroot beforehand.